Feature/dev create brain

JNotelddim opened the pull request:
Created Brain class w/ typings & added all the missing typings for everything brain touches.
Required a lot of test updates.

JNotelddim committed:
Create brain class

brain includes basic move logic.

board gains width property for understanding boundaries & a function for
determining whether or not a coordinate is within bounds.

snake gains a method for to expose whether or not a coordinate is within
its body.

note: pylint is showing any typings in signatures as invalid syntax, so
I'm not currently using the typings until I can figure out how to fix
that.

JNotelddim committed:
WIP: Create brain test and update other tests

Brain still needs to be completed - but I'm unsure as to whether or not
we can continue with the current approach of not using other classes'
methods and stuff. Before we were avoiding it cause the tests are just
intended to test the one class.

Also, I added tests for new board and snake functions.

JNotelddim committed:
Add some new functionality and some TODOs

Added:
*docstrings to Snake class
*todos to Board
*typings to Brain
*docstrings and Brain to Snake

JNotelddim committed:
Use real Snake in Board

have to update board_test still

JNotelddim committed:
Fix brain syntax

had invalid syntax for return type in brain
removed circular dependency from snake to brain
added typings to snake

JNotelddim committed:
Add ALL the typings + finish first pass on brain

* retyped friggin everything.
* got tests passing again.
* got brain tests passing.

JNotelddim committed:
Remove unwated TODO and trim useless test code

JNotelddim committed:
Add some files with same game data

just for reference when I haven't got internet lol

JNotelddim committed:
Merge branch 'feature/dev-create-brain' of github.com:Zullybur/python-snake into feature/dev-create-brain

mattcasiro commented on src/board.py:
```diff
@@ -1,16 +1,49 @@
+    def foods(self) -> List[Coordinate]:
         """Retreive the list of food from the board data."""
-        return self._data['food']
+        if not self._foods or len(self.foods) <= 0:
```
I think you should be consistent here with how you handle `snakes()` since their data types are the same (lists of something). Both should evaluate to `None` at first, and then not `None` afterward. 

- [x] use `if self._foods is None` like in snakes above

JNotelddim replied:
the type checks complain about None not matching the datatype of List[type] so I'll make them all consistent - but I think I need to keep the len check, because for example when in snake.py's constructor, it initializes self._coordinates as [], it's not None but it still needs to go into the conditional which creates the _coordinates data

mattcasiro commented on src/board.py:
```diff
@@ -1,16 +1,49 @@
+
+            self._foods = foods
+        return self._foods
+        #return self._data['food']
```
- [x] Remove outdated code

mattcasiro commented on src/board.py:
```diff
@@ -1,16 +1,49 @@
+        #return self._data['food']
+
+    @property
+    def width(self):
```
- [x] Add return type

mattcasiro commented on src/board.py:
```diff
@@ -1,16 +1,49 @@
+        """Get width of the board -- note: it's a square."""
+        return self._data['width']
+
+    def is_coordinate_in_bounds(self, coordinate):
```
- [x] Add return type

mattcasiro commented on src/board.py:
```diff
@@ -1,16 +1,49 @@
+        is_in_bounds = not is_wall
+        return is_in_bounds
+
+    def get_other_snakes(self, exclude_id):
```
- [x] Add return type

mattcasiro commented on src/board.py:
```diff
@@ -1,16 +1,49 @@
+        is_wall = (coordinate.x == -1 or coordinate.x == self.width
+                   or coordinate.y == -1 or coordinate.y == self.width)
+        is_in_bounds = not is_wall
+        return is_in_bounds
```
- [x] `return not is_wall`

mattcasiro commented on src/board.py:
```diff
@@ -1,16 +1,49 @@
+
+    def get_other_snakes(self, exclude_id):
+        """Get the List of Snakes whose IDs don't match the given ID."""
+        return [snake for snake in self.snakes if snake.id != exclude_id]
```
This is VERY cool.

JNotelddim replied:
Haha yeah I liked it too! 👍

mattcasiro commented on src/board.py:
```diff
@@ -1,16 +1,49 @@
-        return self._data['food']
+        if not self._foods or len(self.foods) <= 0:
+            foods = []
+            for food_data in self._data['food']:
```
- [x] Do that VERY cool thing from below here too.

mattcasiro commented on src/board.py:
```diff
@@ -1,16 +1,49 @@
-        return self._data['snakes']
+        if self._snakes is None:
+            snakes = []
+            for snake_data in self._data['snakes']:
```
- [x] Do that VERY cool thing from below here too.

mattcasiro commented on src/brain.py:
```diff
@@ -0,0 +1,58 @@
+        moves = ["left", "right", "up", "down"]
+        valid_moves = []
+
+        for move in moves:
```
This loop feels inconsistent inside. I think each section should `continue` on its own. Or each section should set an `is_valid` variable. Or you could even just merge all the snake bodies in to a list and say `valid_moves.append(move) if self.board.is_coordinate_in_bounds(move_coordinate) and move_coordinate not in collision_list`. I don't know, I guess there are lots of options, but I just don't love how there's one `continue` in one condition, then one `continue` over two conditions.

- [x] Revisit this to see if there's a cleaner way to implement it

mattcasiro commented on src/brain.py:
```diff
@@ -0,0 +1,58 @@
+            if distance < closest_food[1]:
+                closest_food = (food, distance)
+
+        if closest_food[0] is None:
```
I don't think you can hit this condition because you set a default coordinate initially. Did you intend to default to `closest_food = (None, 9999.0)`?

- [ ] Revisit this condition

JNotelddim replied:
You're right it wasn't possible to hit. I think i just missed taking out checking for None - originally i had it initializes with an empty object which had None values. but then it wasn't agreeing with typings. I think I wanted it returning None for when there isn't any food on the board (if that's even possible), but yeah with it checking 
`if closes_food[1] == 9999.0:"
it fails the tests cause of the invalid return type: "None"
let's discuss this one later today

mattcasiro commented on src/snake.py:
```diff
@@ -1,23 +1,54 @@
-        return self._data['coords']
+    def coordinates(self) -> List[Coordinate]:
+        """Get the List of Coordinates which describes the location of this Snake."""
+        if self._coordinates is not None and len(self._coordinates) > 0:
```
inconsistent to how you did it in Board, see my comments there for reference.

- [x] Refactor to match style in board

mattcasiro commented on src/snake.py:
```diff
@@ -1,23 +1,54 @@
         return self.coordinates[1:]
 
     @property
     def health(self):
```
- [x] Needs return type

mattcasiro commented on test/board_test.py:
```diff
@@ -1,33 +1,63 @@
             "snakes" : snakes,
-            "food"  : foods
+            "food"  : foods,
+            "width" : width
```
The problem with lining up colons like this without IDE supporting the code style is that it's annoying to maintain. For example, I'm pretty sure food's and width's colons are one space off from snake's.

- [x] Don't line up colons, just use `key: value`

mattcasiro commented on test/brain_test.py:
```diff
@@ -0,0 +1,32 @@
+            ((6, 12))
+        ]
+        board = Board({
+            "snakes" : snakes,
```
- [ ] Don't line up colons, just use `key: value`

mattcasiro reviewed (commented):
Looks like a great foundation for Brain! Looking forward to building more logic on top of it 😄

mattcasiro reviewed (changes requested):
Looks like a great foundation for Brain! Looking forward to building all the logic on top of it 😄

JNotelddim committed:
Fix stylistic consistency issues and formatting and clean up code

*As per some solid review from zully:
use list comprehension [x for x in z] in more places where possible.
note, in brain:getValidMoves() it would be nicer to use List unpacking
    ie: collision_coordinates = [*snake.coordinates for snake in
self.board.snakes]
    however that's not allowed in python, so I had to use nested list
comprehension. (same as nested for loops)

*removed some commented code
*finished  off some missed return types
*brain.py handling no food on the board still needs a better solution
*fixed up KVP styling

JNotelddim committed:
Commit coverage file

JNotelddim committed:
Shorten name of coordinate variable in brain

JNotelddim committed:
Use Optional classes where applicable + increase brain tests

*return None from brain.get_nearest_food() when there's no food
*add case to brain_test.py to test ^ case
*small code cleanups based on suggestions

JNotelddim committed:
Commit missed removal of commented code

JNotelddim committed:
Remove len checks from properties initialized as None

JNotelddim committed:
DeMorgan's da heck outta dat brain.get_valid_moves()
